apps/onboarding/src/app/components/ImageWrapper.tsx

```diff
@@ -0,0 +1,35 @@
+import styled from '@emotion/styled';
+import { PropsWithChildren } from 'react';
+
+interface ImageWrapperProps {
+  width: number;
+  height: number;
+}
+
+const Wrapper = styled.div<ImageWrapperProps>`
+  position: relative;
+  margin: 0 auto;
+  overflow: hidden;
+  width: 100%;
+  height: 100%;
+  max-width: ${({ width }) => width}px;
+  aspect-ratio: ${({ width, height }) => width / height};
+
+  & img {
+    object-fit: 'contain';
+  }
+`;
+
+const ImageWrapper = ({
+  width,
+  height,
+  children,
+}: PropsWithChildren<ImageWrapperProps>) => {
+  return (
+    <Wrapper width={width} height={height}>
+      {children}
+    </Wrapper>
+  );
+};
+
+export default ImageWrapper;
```

Comment:
기존 Wrapper의 prop이 width, height를 받고 있는데 실은 내용을 보면 width가 max-width의 역할을 담당하고 있는데 이는 동작을 오해하게 할 수 있을 거 같아.

나는 이런 경우에 css prop을 사용하면 ImageWrapperProps을 무리해서 styled에 끼워맞추지 않을 수 있을 거 같고 더 단순히 코드를 작성할 수 있을 거 같아

css prop을 활용하면 Wrapper라는 알 수 없는 이름을 만들지 않을 수 있어서 좋다고 생각해 사용하는 측에서 Wrapper라는 이름을 보았을 때 저게 div인지 section인지는 가보아야만 알 수 있으니까

어떻게 생각해?

```suggestion
import { css } from '@emotion/react';
import { PropsWithChildren } from 'react';

interface ImageWrapperProps {
  width: number;
  height: number;
}

export default function ImageWrapper = ({
  width,
  height,
  children,
}: PropsWithChildren<ImageWrapperProps>) => (
    <div css={css`
        position: relative;
        margin: 0 auto;
        overflow: hidden;
        width: 100%;
        height: 100%;
        max-width: ${width}px;
        aspect-ratio: ${width / height};
      
        & img {
          object-fit: contain;
        }
    `}>
      {children}
    </div>
    );
```

Reply:
나는 개인적으로 css prop 을 적게 쓰는걸 선호하는 편인데,
기존에 선언된 컴포넌트가 있고, 그걸 수정해야할 때 css props로 넘기는게 맞다고 생각하는데,
애초에 컴포넌트를 새로 생성해야할 때는 저렇게 분리하는 걸 선호해.

지금이야 코드가 짧으니깐 해당 컴포넌트가 어떤 역할을 하고 있은지 한눈에 볼 수 있지만, 코드가 길어지면 가독성 측면에서 이게 어떤 역할을 하는 컴포넌트인지 알기가 어려워지잖아. 특히나 저기에 로직까지 포함된 경우라면 더더욱 분리되어야 한다고 생각하거든. 

지금 이게 Wrapper라는 이름에 대한 align이 맞지 않아서 생긴 이슈라고 생각해서 + 코드가 전체적으로 통일된 형태면 좋겠어서 조금 더 고민해볼게

Reply:
> 코드가 길어지면 가독성 측면에서 이게 어떤 역할을 하는 컴포넌트인지 알기가 어려워지잖아.

나는 이 부분은 동의하기 어려운 거 같아. styled-component를 사용해도 현재처럼 변수의 네이밍이 Wrapper라면 파악하기 어려운 것은 마찬가지라고 생각해 Wrapper에 직접가서 일일이 확인해야 한다면 변수의 장점도, css-in-js의 장점도 누리고 있지 못하고 있는거니까 Wrapper라는 이름이 Wrapper내부가 어떻게 동작할지 정확히 표현해주고 있지 못하다고 생각해서 그래. 오히려 Wrapper내부가 변경되어야 할 때 깊이가 깊어지기 때문에 명시적이지 못하고 유연하지 못한 코드가 아닐까 싶어.

css prop의 기능이 styled-component, emotion 모두 제공해주는 기능인데 styled component뿐만 아니라 적극적으로 사용하는 걸 고민해봐도 좋을 거 같아. 변수가 많아진다는 것은 관리해야 할 포인트도 많아진다는 뜻이니까

```tsx
interface ImageWrapperProps {
  width: number;
  height: number;
}
const Wrapper = styled.div<ImageWrapperProps>`
  position: relative;
  margin: 0 auto;
  overflow: hidden;
  width: 100%;
  height: 100%;
  max-width: ${({ width }) => width}px;
  aspect-ratio: ${({ width, height }) => width / height};
  & img {
    object-fit: 'contain';
  }
`;
```

예를 들어 Wrapper가 ImageWrapperProps의 props를 사용하고 있는데 만약 ImageWrapper의 props가 더 추가되어야 한다면 어떻게 해야할까? 수정해야하는 곳이 두 곳으로 늘텐데 이런 점이 styled방식의 유연하지 못한 부분이라고 생각해. 

> 특히나 저기에 로직까지 포함된 경우라면 더더욱 분리되어야 한다고 생각하거든.

로직까지 포함된 경우라고 말해줬는데 나는 UI에서 style도 로직중 하나라고 생각해. ImageWrapper에서 관리해야 할 포인트가 다른 변수를 참조해서보다는 참조할 변수가 적은 게 더 쉬운 코드라고 생각해서



현재 Wrapper라는 이름이 표현하는 게 내부 스타일링을 유추할 수 업어서 내부를 들여봐야하고 이건 마치 아래의 코드와 같은 상태라고 생각해

```ts
const foo = ({a, b}: SomethingImportantProps) => a + b
const bar = ({a, b}: SomethingImportantProps) => a - b

type SomethingImportantProps = {a: number, b: number}
export const somethingImportant ({a, b}: SomethingImportantProps) => {
  return foo(a, b) + bar(a, b) // foo, bar의 이름이 내부동작을 예측할 수 없어서 foo의 역할과 bar의 구현을 모두 직접 봐야 somethingImportant의 동작이 예상됨. 잘못된 추상화의 예
}

```

이런 경우에는 아래처럼 foo, bar를 풀어서 쓰는 게 이해하기 쉬운 코드이지 않을까 싶어. 정말 추상화가 필요한 타이밍에 추상화하는 게 더 좋다고 생각해
```ts
type SomethingImportantProps = {a: number, b: number}
export const somethingImportant ({a, b}: SomethingImportantProps) => {
  return a + b + a - b // 풀어서 쓰면 somethingImportant에 들어온 사람이 바로 이해할 수 있음
}
```

Reply:
무슨 얘기인지 이해했어! 고마워~

해당 컴포넌트를 안쓰게 되어서 일단 삭제했어! 위에서 해준 얘기들은 다른 코드들에도 적용해서 생각해볼게!